clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/service/servergroup/OracleBMCSServerGroupPersistence.groovy

```diff
@@ -0,0 +1,257 @@
+   */
+  protected List<String> listServerGroupNames(OracleBMCSPersistenceContext ctx) {
+    def result = []
+    def namespace = getStorageNamespaceFor(ctx)
+    ensureStorageBucketFor(ctx)
+
+    try {
+      def rq = ListObjectsRequest.builder().namespaceName(namespace)
+        .bucketName(SERVERGROUP_BUCKET_NAME)
+        .build()
+      def rs = ctx.creds.objectStorageClient.listObjects(rq)
+      def maybeListSummaries = rs?.getListObjects()?.getObjects()
+      if (maybeListSummaries != null) {
+        result = maybeListSummaries.collect { it.getName() }
+      }
+    }
+    catch (Exception e) {
+      throw new OracleBMCSServerGroupPersistenceException("Failed listing contents of bucket", e)
+    }
+
+    return result
+  }
+
+  /**
+   * Reads the server group data with the provided name in the provided account.
+   */
+  OracleBMCSServerGroup getServerGroupByName(OracleBMCSPersistenceContext ctx, String name) {
+    return doPersistenceOperation(PersistenceOperation.READ, ctx, null, name)
+  }
+
+  /**
+   * Writes the server group data to the persistent store; the account is inferred from the server group data.
+   */
+  void upsertServerGroup(OracleBMCSServerGroup sg) {
+    if (sg.credentials.compartmentId == sg.launchConfig["compartmentId"] as String) {
+      doPersistenceOperation(PersistenceOperation.UPSERT, new OracleBMCSPersistenceContext(sg.credentials), sg, sg.name)
+    } else {
+      throw new OracleBMCSServerGroupPersistenceException("Different compartments - this is not allowed")
+    }
+  }
+
+  /**
+   * Deletes the server group data from the persistent store; the account is inferred from the server group data.
+   */
+  void deleteServerGroup(OracleBMCSServerGroup sg) {
+    if (sg.credentials.compartmentId == sg.launchConfig["compartmentId"] as String) {
+      doPersistenceOperation(PersistenceOperation.DELETE, new OracleBMCSPersistenceContext(sg.credentials), sg, sg.name)
+    } else {
+      throw new OracleBMCSServerGroupPersistenceException("Different compartments - this is not allowed")
+    }
+  }
+
+  @Synchronized
+  private String getStorageNamespaceFor(OracleBMCSPersistenceContext ctx) {
+    if (ctx.namespace) {
+      return ctx.namespace
+    }
+    def rq = GetNamespaceRequest.builder().build()
+    def rs = ctx.creds.objectStorageClient.getNamespace(rq)
+    if (rs?.getValue() == null) {
+      throw new OracleBMCSServerGroupPersistenceException("Namespace not found, can't continue")
+    }
+    ctx.namespace = rs.getValue()
+    return rs.getValue()
+  }
+
+  @Synchronized
+  private ensureStorageBucketFor(OracleBMCSPersistenceContext ctx) {
+    if (ctx.bucketChecked) {
+      return
+    }
+    try {
+      def rq = HeadBucketRequest.builder().namespaceName(ctx.namespace).bucketName(SERVERGROUP_BUCKET_NAME).build()
+      def rs = ctx.creds.objectStorageClient.headBucket(rq)
+      if (rs?.getETag() != null) {
```

Comment:
Can `eTag` be `""`? If so, and if that should also be guarded here, maybe `if (rs?.getETag())`?